### ISSUE TITLE
Problem with custom plugins when using git-sync container

### ISSUE THREAD

**Issue by rbakhtaraev** [apache/airflow#14714]
**Apache Airflow version**: 1.10.14/2.0.0 (it isnt important)

**Kubernetes version (if you are using kubernetes)** (use `kubectl version`): 1.20.1

**Environment**:

- **Cloud provider or hardware configuration**: 
- **OS** (e.g. from /etc/os-release): 
- **Kernel** (e.g. `uname -a`): 
- **Install tools**: helm with using chart from this repo
- **Others**: celery executor with 1 worker

**What happened**:

Custom plugins don't work, when using git-sync container for fetch their from git. 
I think the issue may be in a fact that plugins and dags are fetching later than worker is starting.
If i add some pause before starting worker/webserver through `sleep 30`, worker/webserver is starting postponed and plugins are loading.

Is there a better solution for my case?

**Example**:
```
Broken DAG: [/opt/airflow/dags/repo/dags/youtrack_history.py] Traceback (most recent call last):
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/opt/airflow/dags/repo/dags/my-custom-dag.py", line 2, in <module>
    from my-custom-module import CustomClass
ModuleNotFoundError: No module named 'my-custom-module'
```
When i try to execute this dag from pod (`k exec -it pod-id -c webserver bash -c python ./.../my-dag`) , everything is ok.

Adding `sleep 30` into entrypoint-script solve this problem.

**Comment by mik-laj** [comment]
@rbakhtaraev You can add an init container that will fetch the repository initially.

**Comment by rbakhtaraev** [comment]
> @rbakhtaraev You can add an init container that will fetch the repository initially.

It's requires a support of chart locally (e.g. custom deployment.yaml), but that isn't convenient.

Can we add it into chart? I'm ready to send PR with edits.

How i see that:
We have an init-container with shared dir, which fetch a repo, after that shared dir passing into airflow and git-sync container and update continuously

**Comment by mik-laj** [comment]
> Can we add it into chart? I'm ready to send PR with edits.

Yes. It will be helpful. Happy to help with review.

**Comment by kaxil** [comment]
Hi @rbakhtaraev -- what were the things you were importing via plugins.

Just wanted to check since you can import Operators, sensors etc without plugins

**Comment by rbakhtaraev** [comment]
> Hi @rbakhtaraev -- what were the things you were importing via plugins.
> 
> Just wanted to check since you can import Operators, sensors etc without plugins

Hi @kaxil , im using plugins to some custom clients for internal tools. For example, i have 5 dags, which imported two custom plugins.

**Comment by kaxil** [comment]
Can you give an example code, please use normal python modules or package, check https://airflow.apache.org/docs/apache-airflow/stable/modules_management.html

**Comment by yesemsanthoshkumar** [comment]
I faced the similar issue as well. The custom operators wouldn't load even if they are inside the dags folder. Tried loading it from the plugins dir as well.

What worked for me is setting `reload_on_plugin_change` to `True` in the values.yaml. Operators got picked up from the dags folder itself.

 ````
 from datetime import datetime

from airflow.models import DAG
from airflow.operators.dummy import DummyOperator

from custom_operator import MyDummyOperator

default_args = {
    'owner': 'airflow,
    'start_date': datetime(2021, 6, 1),
    'depends_on_past': False,
}

dag = DAG(
    dag_id='import_test',
    schedule_interval=None,
    default_args=default_args,
    catchup=False
)

op = DummyOperator(
    task_id='dummy',
    dag=dag
)

my_op = MyDummyOperator(
    task_id='dummy_mine',
    dag=dag
)
````

and my custom operator was 

```
from airflow.operators.dummy import DummyOperator

class MyDummyOperator(DummyOperator):
    pass
```

I also found that this works even though I don't have the plugins folder and all the imports are within the dags folder. Which didn't workout when setting it to false. @kaxil Am I on the right path?

**Comment by kaxil** [comment]
>I also found that this works even though I don't have the plugins folder and all the imports are within the dags folder. Which didn't workout when setting it to false. @kaxil Am I configuring it correctly? I thought this reload was intended for the plugins folder.

That does not feel right. The operator is unrelated to any sort of plugin.

**Comment by KIRY4** [comment]
Hello @kaxil ! We have a similar problem. But there are some differences. We are using Azure Kubernetes Service and for dags logs plugins and our custom_operators we are using Kubernetes PVC's with Azure file shares under the hood. Also we are using community Helm chart: https://github.com/airflow-helm/charts/tree/main/charts/airflow

After uploading custom operator on mounted as PV/PVC Azure file shares I expect to have it accessible from Airflow DAGs.
I have 2 folders: /opt/airflow/plugins and /opt/airflow/custom_operator.

Examples of imports (for test I jumped into Web POD):

From plugins/ I can call custom opertor like - from table_migration_operator import TableMigrationOperator
From custom_operator/ - from custom_operator.save_result_operator import SaveResultOperator
<img width="1484" alt="image" src="https://user-images.githubusercontent.com/42247943/125605632-82e702f0-92c7-42c5-8752-cd66b9d22f88.png">
Inside POD from Python everything works fine.


What I'm getting in UI:
<img width="642" alt="image" src="https://user-images.githubusercontent.com/42247943/125605859-a6c170c9-c61d-4d29-9514-e32a9b2c9d42.png">



 

Both ways are not working in UI until webserver will not be restarted (restart pod with webserver). After the restart everything is fine.

Airflow versions: 2.1.0 and 2.1.1

PATH=/home/airflow/.local/bin:/usr/local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin

PYTHONPATH=/opt/airflow:/opt/airflow/dags:/opt/airflow/logs:/opt/airflow/plugins:/opt/airflow/custom_operator:/opt/airflow/jars

AIRFLOW__WEBSERVER__RELOAD_ON_PLUGIN_CHANGE=True

**Comment by potiuk** [comment]
Just to clarify (we recently agreed on naming) - this is the "User-community" chart :) - I hope my PR https://github.com/airflow-helm/charts/pull/294 will be merged soon and we will get rid of the ambiguity.

If you want to continue with Matthew's (User-community) chart you should raise the issue there  https://github.com/airflow-helm/charts/  - we do not directly support the non-official chart here, but we do offer help if you want to migrate to the official one - you can open a discussion or start discussing in slack (we just created `#helm-chart-official` channel there.
